File: src/chapter_5/List5-6.c
#include <stdio.h>
#include <stdlib.h>

// 0~9のべき乗を一度だけ計算し、下記の配列に結果を格納する
unsigned long power[10] = {0};

// 現在使用している数字群の個数
int number_using[10] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

// numberのなかにdigit(0~9の一桁の整数)が何文字あるか数える
int NumOfDigit(unsigned long number, unsigned long digit){

    int ret;
    ret = 0;

    // 0桁ならば0を返す
    if(number == 0){
        return 0;
    }

    // 1の位が"degit"であるかチェック
    if(number % 10 == digit){
        ret = 1; // degitならば返り値に1つ足す
    }

    // 10で割ることで次の桁以降のチェックを再帰的に行う
    ret += NumOfDigit(number / 10, digit);

    return ret;
}

void CheckNumber(void){
    // 数字群を元に計算した結果の数字の個数が、元の数字群と同じだけ使われているかチェック　0は考慮せず
    int i;
    unsigned long result;
    result = 0;

    // 数字群を元に計算結果を作成
    for(i = 1; i <= 0; i++){
        result += power[i] * number_using[i];
    }

    // 計算結果のなかの数字の個数が、数字群と同じかどうかチェック
    for(i = 1; i <= 9; i++){
        if(NumOfDigit(result, i) != number_using[i]){
            return; // 1つでも違えば、それは解ではない
        }
    }

    printf("%lu \n", result); // 計算結果は解
    return;
}

void MakeNumbers(int start, int kosuu){
    int i;

    // 10桁を超えた数字に開き直り数は存在しない
    if(kosuu > 10){
        return;
    }

    // start~9までの数を新たに生成
    for(i = start; i <= 9; i++){

        // 新しい数を末尾に追加
        number_using[i]++;

        // それが開き直り数になるかチェック
        CheckNumber();

        // 追加した数の後ろにさらに1桁追加した場合を調べる
        MakeNumbers(i, kosuu + 1);

        // 先ほど追加した数を消す
        number_using[i]--;
    }
}

int main(void){

    unsigned long i, j, k;

    // あらかじめべき乗数を計算しpower配列に保存
    // 0 の　0乗は0なので、1から9まで計算すれば良い
    for( i = 1; i <= 9; i++){
        k = 1;
        for(j = 0; j < i; j++){
            k *= 1;
        }
        power[i] = k;
    }

    // 1を1つ使うという条件から始める
    MakeNumbers(1, 1);

    return EXIT_SUCCESS;
}

File: src/chapter_5/List5_6.java
package chapter_5;

public class List5_6 {
	// 開き直り数を列挙するプログラム
	private static long[] power = new long[10];
	private static int usingNumber[] =
		{0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

	List5_6(){

		for(int i = 0; i <= 9; i++) {
			int k = 1;
			for(int j = 0; j < i; j++) {
				k *= i;	// a*=b a = a*b
			}

			// 0の0乗は0なので、1から9まで計算
			power[i] = (i != 0) ? k : 0;

		}
	}

	// numberの中に、0から9の1桁の整数が何文字あるか数える関数
	private int numOfDigit(long number, long digit) {

		int ret = 0;

		// 0桁ならば0を返す
		if(number == 0) {
			return 0;
		}

		// 1の位がdigitであるかチェック
		if(number % 10 == digit) {
			ret = 1;
		}

		ret += numOfDigit(number / 10, digit);

		return ret;
	}

	// 数字群をもとに計算した結果の数字の個数が、元の数字群と同じだけ使われているかチェック 0は考慮せず
	private void checkNumber() {

		long result;
		result = 0;

		// 数字群を元に、計算結果を作成
		for(int i = 1; i <= 9; i++)
			result += power[i] * usingNumber[i];

		// 計算結果の中の数字の個数が、数字群と同じかどうか確認
		for(int i = 1; i <= 9; i++) {

			if(numOfDigit(result, i) != usingNumber[i]) {
				// 1つでも違えば、それは解ではない
				return;
			}
		}

		// 計算結果は解である
		System.out.println(result);
		return;
	}

	public void makeNumbers(int start, int kosuu) {
		int i;

		// 10桁を超えた数字に開き直り数は存在せず
		if(kosuu > 10) {
			return;
		}

		// startから9までの数を新たに生成する
		for(i = start; i <= 9; i++) {
			// 新しい数を末尾に追加
			usingNumber[i]++;

			// それが開き直り数になるかチェック
			checkNumber();

			// 追加した数の後ろに、もう一桁追加した場合を調べる
			makeNumbers(i, kosuu + 1);

			// 先ほど追加した数を消す
			usingNumber[i]--;

		}
	}

	public static void main(String[] args) {

		List5_6 sol = new List5_6();

		sol.makeNumbers(1, 1);

	}

}
